Splice Plugin: expands TH splices and QuasiQuotes

konn opened the pull request:
This plugin provides an ability to expand splices in the source code.
This is particularly useful when:

1. TH macro produces invalid (e.g. type-error) code, but the output AST is relatively complex and you want to investigate the generated code in detail.
    * Of course, passing `-ddump-splices` just helps in simple case; but it omits some parts when the output size is relatively large.
2. You want to get rid of TH code for some reasons (e.g. Perhaps for cross-compilation?)
    * This use-case is, however, primarily out of the scope of this PR at this stage, as in several cases just dumping generated code doesn't give us the working code:
        1. Hygiene problem: unique flavour of `Name` sometimes get in the way so that `x` and `x_awa2221` won't much as a plain source code
        2. TH macro often includes non-exported symbols (it is a customary pattern to hide an implementation detail to the users and let macro take care of them), hence generated code doesn't have access to them.

By the way, I really need the feature for the reason 1, and I'm pretty sure that it must help us a lot when debugging complex macros.

## Demo (added afterwards)

https://user-images.githubusercontent.com/21860/103377752-9fb04500-4b23-11eb-9121-2d73f7288dd4.mov



## Some notes

### `Ide.TreeTransform`
I really loved the functionality provided in `Ide.TreeTransform`, which was provided in Tactic Plugin.
I wanted to use it also in this plugin, so I extract the module into the new package.
If there is any problem, please tell me @isovector @WorldSEnder

### Non-invasive expansion
The current implementation just replaces the splice with the generated code.
However, it might be sometimes the case that one wants to retain the original splice and just append generated code as the comment right after the splice.
I once tried to use `ghc-exactprint`, but couldn't work out the way to insert new comments with it.

### Declaration Splice
It seems HieAST doesn't contain Declaration Splices at all, and contrary to the rest of ASTs, they are one-to-many correspondence with the original splice.
Hence, I just skip the implementation of the expansion of declarations. Perhaps worth another PR?

## TODOs:

- [x] Implement golden tests
- [x] Support declaration splices
- [ ] Support commented-out style of expansion; i.e. not replacing the old splice, but expands them surrounded by the comment braces, right after the splice.

## Credits
This pull request also incorporating splice hover support, which was at first implemented in [mpickering/ghcide/top-level-splices-hover](https://github.com/mpickering/ghcide/tree/wip/top-level-splices-hover).
Thank you @mpickering to allow me to include it and  thank you @wz1000 for suggesting that!

wz1000 commented on plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs:
```diff
@@ -0,0 +1,364 @@
+                            }
+                    env' <- getSession
+                    -- setTargets [thisModuleTarget]
+                    resl <- load LoadAllTargets
```
Why do this? `load` will effectively double IDE memory usage.

konn replied:
At first, I couldn't make out how to make symbols in splice into Renamer's scope and I struggled to resolve such situations just by stealing some codes from Eval Plugin.
At last, I found that these and passing `TcGblEnv` to `init*Tc` resolves things, but, yes, I must definitely try to remove some/all fragments of these HscEnv setting hackeries. I'll try, Thanks!

konn replied:
It seems we don't need `evalGhcWith` hacks isn't needed entirely, so I deleted that part. Thanks again!

wz1000 commented:
You can use `runMetaHook` to directly capture the result of a splice when ghc is running it: https://hackage.haskell.org/package/ghc-8.10.1/docs/Hooks.html#v:runMetaHook

konn commented:
> You can use `runMetaHook` to directly capture the result of a splice when ghc is running it: https://hackage.haskell.org/package/ghc-8.10.1/docs/Hooks.html#v:runMetaHook

Thanks for pointing it out. I wasn't aware of it. I'll give it a try!

wz1000 commented:
See this for an example of how to set and use a MetaHook: https://github.com/mpickering/ghcide/commit/d9e59abc8616284fa900d7f80324c725077929bb

In this code, the input to the splice is being captured - you would like to capture the output instead.

wz1000 commented on plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs:
```diff
@@ -0,0 +1,334 @@
+    type SpliceOf ast :: Kinds.Type -> Kinds.Type
+    type SpliceOf ast = HsSplice
+    matchSplice :: Proxy# ast -> ast GhcPs -> Maybe (SpliceOf ast GhcPs)
+    renameSplice :: Proxy# ast -> SpliceOf ast GhcPs -> RnM (Either (ast GhcPs) (ast GhcRn), FreeVars)
```
Do we need to rename the splice? Isn't the parsed version good enough for pretty printing?

konn replied:
Yes. But, AFAIK, GHC API only provides splice expansion functions with result type in either `GhcRn` or `GhcTc` phases (provided in `RnSplice` and `TcSplice` resp).
Since splicing is taken place in renaming ang typing phase, this design seems reasonable.
Only exception was Pat, as `rnSplicePat` returns `Either GhcPs GhcRn`.

I used `rnSplice*` family provided in `RnSplice` to expand splices, so I named corresponding method as  `renameSplice`, but considering what it actually do, perhaps I should have named it `expandSplice`.

wz1000 replied:
OK. The MetaHook should allow you to access the parsed representation of the splice, also avoiding any overhead from having to re-run splices.

konn replied:
That sounds great! It seems I should definitely give a try to `MetaHook`. Thanks!

konn commented:
Hmm, as far as I tried, it seems that `MetaHook` combined with `hscTypecheckRename` works only when there is no type-errors:

```hs
captureSplices ::
    HscEnvEq ->
    ParsedModule ->
    (HscEnv -> ExpandedSplices -> MaybeT IO a) ->
    MaybeT IO a
captureSplices hscEnvEq pm k = do
    var <- liftIO $ newIORef (mempty :: ExpandedSplices)
    let hsEnv0 = hscEnv hscEnvEq
        dflags0 = hsc_dflags hsEnv0
        hooks0 = hooks dflags0
        dflags = dflags0 {hooks = hooks0 {runMetaHook = Just $ mkSpliceHook var}}
        hsEnv = hsEnv0 {hsc_dflags = dflags}
    void $
        liftIO $
            hscTypecheckRename
                hsEnv
                (pm_mod_summary pm)
                HsParsedModule {hpm_module = parsedSource pm, hpm_src_files = pm_extra_src_files pm, hpm_annotations = pm_annotations pm}
    k hsEnv =<< liftIO (readIORef var)
```

As my first motivation is to debug erroneous macros, it is not desirable such behaviour... Is there any working workaround?

wz1000 commented:
Did you try setting `-fdefer-type-errors`? Also, please add this hook into the usual ghcide typecheck pipeline, we really don't want to be typechecking modules twice (not only is this wasteful, it is unsafe in the presence of Template Haskell, see https://gitlab.haskell.org/ghc/ghc/-/issues/18812).

The `ghcide` typechecking logic (in `ghcide/src/Development/IDE/Core/Compile.hs`) already has `-fdefer-type-errors` set, so if you just stick this logic in there (possibly adding another field to the result of typechecking to capture the splices), everything should just work.

This is easy now that we have merged repositories.

konn commented:
> Did you try setting `-fdefer-type-errors`? Also, please add this hook into the usual ghcide typecheck pipeline, we really don't want to be typechecking modules twice (not only is this wasteful, it is unsafe in the presence of Template Haskell, see https://gitlab.haskell.org/ghc/ghc/-/issues/18812).
> 
> The `ghcide` typechecking logic (in `ghcide/src/Development/IDE/Core/Compile.hs`) already has `-fdefer-type-errors` set, so if you just stick this logic in there (possibly adding another field to the result of typechecking to capture the splices), everything should just work.

Ah! I forgot about `-fdefer-type-errors` and didn't know that they are already included in ghcide. I'll try it!
If I have to extend ghcide, then I can just add the field corresponding to ExpandedSplcies in TcModuleResult, right?

wz1000 commented:
While you are doing this, maybe you can even pick up @mpickering's patch for https://github.com/haskell/haskell-language-server/issues/733 ?

The branch is linked in the issue and the implementation should be mostly complete. The same hook can be used to capture all the information required.

wz1000 commented:
>  If I have to extend ghcide, then I can just add the field corresponding to ExpandedSplcies in TcModuleResult, right?

Yes, please do.

konn commented:
OK, I confirmed that setting `-fdefer-type-errors` makes it possible to expand erroneous splices. I'll start to integrate splice capturing into ghcide!

wz1000 commented on ghcide/src/Development/IDE/Core/RuleTypes.hs:
```diff
@@ -90,13 +93,38 @@ newtype ImportMap = ImportMap
+
+instance NFData SpliceInfo where
+  rnf SpliceInfo{..} =
+     exprSplices `seq` patSplices `seq` typeSplices `seq` declSplices `seq` ()
```
Use `rnf` to fully evaluate this.

konn replied:
`HsExpr` and his friends don't come with NFData instance, so I chose to reduce them just to WHNF, as it seems there already exist such cases just reduces fields just to whnf. Are there any wrappers to generically traverse them?

wz1000 replied:
You can add `NFData` instances for `HsExpr` etc. in the Orphans module. The implementation for those can just be `seq`. That way, at least the spine of the list is forced. Alternatively, you can use `liftRnf seq`.

wz1000 commented on ghcide/src/Development/IDE/Core/Compile.hs:
```diff
@@ -149,21 +150,49 @@ typecheckModule (IdeDefer defer) hsc keep_lbls pm = do
+    splice_hook :: IORef SpliceInfo -> MetaHook TcM
+    splice_hook var metaReq e = case metaReq of
+        (MetaE f) -> do
+            expr' <- metaRequestE defaultRunMeta e
```
You should use the original hook instead of `defaultRunMeta` so that hooks set elsewhere aren't silently overwritten.

konn replied:
Done. Thanks for pointing out!

wz1000 commented on ghcide/src/Development/IDE/Core/RuleTypes.hs:
```diff
@@ -1,48 +1,49 @@
+{-# LANGUAGE DerivingStrategies #-}
 -- Copyright (c) 2019 The DAML Authors. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
-
```
Please don't include unrelated formatting changes.

konn replied:
Thanks for pointing it out. I tried hard not to include them (using Save without Formatting), but as I do it manually, there can be such changes by accident. I'm really longing to see #693 to be resolved...

wz1000 commented on plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs:
I think you don't need the `HieAST` now, you can just look at the source spans of the original unexpanded splices.

konn replied:
Done! There is a pitfall that HieAST and ParsedModule gives slightly different Locations for splices, resulting in the change of searching strategy for splice span. As an additional benefit, we can now invoke the CodeAction over whole splices; formerly we can invoke CodeLens only strictly inside splices (e.g. right after `(` from before `)` in `$(foo)`).

konn commented:
It seems type-level splices with kind error cannot be expanded. Perhaps, we could fallback to the old implementation when type-level splice information on the span isn't available from TcModuleResult.

wz1000 commented:
> Perhaps, we could fallback to the old implementation when type-level splice information on the span isn't available from TcModuleResult

Sounds reasonable, though I'm not sure how you are getting this to work. If the code fails during kind-checking, how are you getting any information about the splice out?

mpickering commented:
Thanks for incorporating my patch into this and thanks to @wz1000 for pointing it out.

mpickering commented:
Did you deliberately only take the first commit from the branch? There are two more you need as well.

konn commented:
Oh, I overlooked other commits. I will cherry-pick them!

konn commented:
@mpickering I've just cherry-picked two more commits and adjusted to new interfaces. Thanks!

mpickering commented on plugins/hls-splice-plugin/hls-splice-plugin.cabal:
```diff
@@ -27,6 +27,7 @@ library
                   , dlist
                   , retrie
                   , shake
+                  , syb
```
Using `syb` is going to be probably very slow without care.

konn replied:
Yeah, but it is rather clumsy and fragile to GHC API change to pattern matching on entire `HsExpr`, `HsPat` and `HsType`, so some kind of use of `syb` should be reasonable.

konn replied:
I tried not to traverse out-of-scope branches in https://github.com/haskell/haskell-language-server/pull/759/commits/f2efe7cc1d0a10e84eec00a684253559ad2bde44. Hopefully this should speed up the usage of syb (though there have been no noticable slow down introduced by syb so far)

wz1000 commented on plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs:
```diff
@@ -198,45 +208,53 @@ unRenamedE dflags expr = do
+                    rangeToRealSrcSpan ran $
+                        fromString $
+                            fromNormalizedFilePath fp
+                mouterSplice = something' (detectSplice spn) pm_parsed_source
```
Why not just check if `spn` is inside one of the spans of the unexpanded splices? (i.e. `spn` lies within one of the Locateds of the `LHsExpr GhcTc`)

konn replied:
Because there can be nested splices in practice and hence we must choose the nnermost one for disambiguation:

```hs
a x = putStrLn $([|"foo" + $(varE $ mkName "x")|])
```

wz1000 replied:
You can pick the smallest one from the list of `LHsExpr GhcTc` that fits. I don't see the need for a syb traversal when we've already captured all the information we need.

konn replied:
For the splices that passed kind-checking, using TypeCheck result makes sense.
But, if one wants to detect all the splice, we need to analyse through the AST.
I confirmed this by experimenting by feeding the following:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TemplateHaskell #-}
module TTypeKindError where
import Language.Haskell.TH ( numTyLit, litT )
import Data.Proxy ( Proxy )

main :: $(litT (numTyLit 42))
main = return ()
```

Using raw `ParsedModule`, `$(litT ...)` part will be detected. But when using `tmrToplevelSplices`, it won't be detected.

konn replied:
Another concern to made me using `ParsedModule` here was that type-checking could take much longer than just parsing, so I think it might be faster to use in CodeAction detection phase than requesting `TypeCheck`.

wz1000 replied:
How will you expand it if it is not detected?

Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

> so I think it might be faster to use in CodeAction detection phase than requesting TypeCheck

If you `useWithStale`, this will not be a concern in 99% of cases

konn replied:
> How will you expand it if it is not detected?

In some cases, my first implementation should work. For example, the above example won't kind-check as a whole, but the splice `$(litT (numTyLit 42))` itself can be expanded if it is considered standalone; and users should expect them to be expanded when debugging.
We can apply functions from `RnSplice` to such examples.

> Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

Yes, it might be helpful to store expanded results in `ExpandSpliceParams`, perhaps wrapped in Maybe.

> If you useWithStale, this will not be a concern in 99% of cases

Sounds good. How likely is it possible that staled module could differ from the current state and destroys the source entirely?
During developing this plugin, I've encountered the situation that Tactic Plugin destroys some lines.

wz1000 replied:
> Sounds good. How likely is it possible that staled module could differ from the current state and destroys the source entirely?

If you `useWithStale` to generate the code action suggestions and then do proper `use` to generate the actual edit, I don't think we will ever mangle the source file.

konn replied:
I decided to keep using ParsedModule in splice detection, since, as mentioned in https://github.com/haskell/haskell-language-server/pull/759#issuecomment-753313091, now we can expand some splices faild due to type- or syntactic-error given that staled TcModuleResult is available.

> Also, you can probably save and return the splice information even if the module fails to typecheck - this will require some wrangling though to figure out where and how to return this information.

I will try this approach in another p.r., if we can find the way to achieve this!

konn commented:
Whoops, I accidentally committed intermediate code that contains partial cases which lacks `HsDecl` case.

konn commented:
A prototype of declaration splicing has been implemented, but it seems works improperly as it deletes all the preceding comments and blanks.

For example:

```hs
{-# LANGUAGE TemplateHaskell #-}
module TSimpleDecl where
import Language.Haskell.TH ( mkName, clause, normalB, funD, sigD )

-- Foo
--  Bar
$(sequence
    [sigD (mkName "foo") [t|Int|]
    ,funD (mkName "foo") [clause [] (normalB [|42|]) []]]
    )
```

becomes:

```hs
{-# LANGUAGE TemplateHaskell #-}
module TSimpleDecl where
import Language.Haskell.TH ( mkName, clause, normalB, funD, sigD )
foo :: Int
foo = 42
```

Perhaps I misused the ghc-exactprint and misimplemented `graftMany`...

konn commented:
OK, I added a dirty hack to adjust starting line of Declaration splice expansion, making sure Decl splice expands as expected.

https://user-images.githubusercontent.com/21860/103438853-16ac2180-4c7b-11eb-9050-c20f4c0d209d.mov

In addition, I've just added some workaround for splice expansion with a kind error.
It just applies the old splice expansion algorithm using `RnSplice`API.
It needs `TcGblEnv`, so we retrieve from staled `TypeCheck` action result, so it fails when the file had kind errors when it was opened.
In such a case, one can manually replace erroneous macros with valid simple code, let HLS to type-check, and then revive the macro and let HLS to expand.

DEMO:


https://user-images.githubusercontent.com/21860/103438930-d6996e80-4c7b-11eb-9f08-448d700330fe.mov

konn commented:
I think I already implemented all the features what I anticipated at first, except for commented-out expansions.
It should take for a while to investigate the way, I think the current features in PR could be merged after I implement regression tests for erroneous splices and the review and CI are green.

konn commented:
Aaaaagggghhhhhhh! In GHC 8.6 and 8.10, `LPat` is isomorphic to `Located Pat`, but in 8.8, it is just `Pat`...

konn commented:
At least it compiles with GHC 8.8, but since the location information was lost in `LPat` in GHC 8.8, pattern splice expansion stops working on GHC 8.8...

konn commented:
It seems matching on `XPat` only in GHC 8.8 almost works, but it seems that it gives slightly different location.

konn commented:
Fmm, it seems that the workaround deviced for `LPat = Pat` problem, then it breaks build on 8.10.
It makes me almost crazy, so I decided not to support pattern splice expansion on GHC 8.8 for a while.

konn commented:
Hmm. Some tests are failing due to Timeout, despite at least for splices related tests are passing at my environment...

konn commented:
I hope that the failures are nonessential, and resolve Draft state. Let's see if it passes after merging.

konn commented:
Hmm... From the error log, it seems the client receives the response for different inputs:

<details>
<summary>Excerpted from 8.8 test error logs</summary>

```
    TQQExp.hs (golden):                                                                                 FAIL
483
      Exception: Timed out waiting to receive a message from the server.
484
      Last message received:
485
      {
486
          "tag": "NotPublishDiagnostics",
487
          "contents": {
488
              "jsonrpc": "2.0",
489
              "params": {
490
                  "uri": "file:///home/runner/work/haskell-language-server/haskell-language-server/test/testdata/CodeActionRename.hs",
491
                  "diagnostics": [
492
                      {
493
                          "severity": 1,
494
                          "range": {
495
                              "start": {
496
                                  "line": 0,
497
                                  "character": 7
498
                              },
499
                              "end": {
500
                                  "line": 0,
601
                          "range": {
602
                              "start": {
603
                                  "line": 0,
604
                                  "character": 7
605
                              },
606
                              "end": {
607
                                  "line": 0,
608
                                  "character": 15
609
                              }
610
                          },
611
                          "source": "typecheck",
612
                          "message": "â¢ Variable not in scope: butStrLn :: [Char] -> t\nâ¢ Perhaps you meant âputStrLnâ (imported from Prelude)"
613
                      },
614
                      {
615
                          "severity": 1,
616
                          "range": {
617
                              "start": {
618
                                  "line": 1,
619
                                  "character": 6
620
                              },
621
                              "end": {
622
                                  "line": 1,
623
                                  "character": 13
624
                              }
625
                          },
626
                          "source": "typecheck",
627
                          "message": "â¢ Variable not in scope: putStrn :: [Char] -> t\nâ¢ Perhaps you meant one of these:\n    âputStrâ (imported from Prelude),\n    âputStrLnâ (imported from Prelude)"
628
                      }
629
                  ]
630
              },
631
              "method": "textDocument/publishDiagnostics"
632
          }
633
      }
634
```
</details>

I  have no idea what's going on here since it works just fine when run on my environment as standalone...

konn commented:
Hmm... ran on my environment, all the splice test passes, but Eval plugin tests started to fail due to `Received an exepected error in a response for id IdRspInt 1`.

konn commented:
It seems that, in my environment, Eval plugin tests fail also on master branch, although CI is passing for master. Perhaps I misunderstand how to run tests?

I ran following:

```sh
$ stack test haskell-language-server:test:func-test --ta "-p eval -j1"
```

## Environment
- macOS Catalina 10.15.7
- compiled with stack.yaml
- branch `master`

konn commented:
I realised that the log told that `QQ.hs` not found. I've just added hie.yaml including QQ.hs, hopefully, it should now compile.

konn commented:
It seems newly added stack-8.10.3.yaml lacked my plugin-related packages. Now that I added it and GitHub Actions are all green, I think this should fix all CI.

@wz1000 @mpickering Thank you for your help so far! Are there any other concerns or possible fixes on this PR?

konn commented:
I found that I forgot to modify DynFlags in default strategy. It must now respects LANGUAGE pragmas in module when parsing-unparsing inside ghc-exactprint.

konn commented:
...And it seems that CI fails unexpectedly due to the current change in the master branch? I will file another issue.

wz1000 commented on plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs:
You can use `(dL -> L l@(RealSrcSpan spLoc) pat :: Located (Pat GhcPs))` for 8.8

https://hackage.haskell.org/package/ghc-8.8.3/docs/GHC.html#v:dL

konn replied:
Oh, I wasn't aware of that! Last time, I tried to extract location info from `XPat`, constructor, which gave a slightly different SrcLoc than expected. Contrary, at least on my laptop, it seems that the combination with `cL` and `dL` gives a  correct SrcSpan and Pattern Splice Expansion successes also in GHC 8.8!

wz1000 reviewed (approved):
Looks good. Seems like you put in a lot of work into making sure this is robust in the face of errors!

konn commented:
@wz1000 Thanks! I owe you and @mpickering for many suggested improvements!

konn commented:
I think the above last commit that enables pattern splice expansion GHC 8.8 is the last change I would add to this PR and it can now be merged when the CI is all green. Thanks!

jneira commented:
nice, many thanks for this new plugin

isovector commented on hls-exactprint-utils/src/Ide/TreeTransform.hs:
```diff
@@ -0,0 +1,334 @@
+        )
+        a
+
+graftWithSmallestM ::
```
@konn can you walk me through what this function is supposed to do? My understanding based on the name is that it should only run the transformation on the smallest `Located ast` that contains the span, but I don't see how that could be the case based on the implementation.

konn replied:
> can you walk me through what this function is supposed to do?

Sure!

> My understanding based on the name is that it should only run the transformation on the smallest Located ast that contains the span

That's what I originally intended, but not used at last. Here, I used `everywhereM'`, defined [Line 266](https://github.com/haskell/haskell-language-server/pull/759/files/6341b396186a7411fc860e654e63f8559611be7e#diff-a278397f41e59b2252d3f37bbb1795b8758f77305fe719c1bd83de3aa24f5711R266). which is a bottom-up variant of `everywhereM` from `syb`.
Since it is bottom-up, I once expected it rewrites the smallest AST containing the given span.
But now, I think this was mis-implemented: it MUST use bottom-up variant of `somewhere` instead of `everywhereM'`, otherwise it will transform ALL the AST element containing the span.

isovector replied:
Mind taking a look at my changes in #1543?

konn replied:
OK, I'll take a look in this evening (in JST)!

isovector commented on hls-exactprint-utils/src/Ide/TreeTransform.hs:
```diff
@@ -0,0 +1,334 @@
+    SrcSpan ->
+    (LHsDecl GhcPs -> TransformT m (Maybe [LHsDecl GhcPs])) ->
+    Graft m a
+graftDeclsWithM dst toDecls = Graft $ \dflags a -> do
```
@konn I also don't understand what this function is doing. Mind walking me through it? Especially what `go` does.

konn replied:
The intended usage of `graftDeclsWithM` is to apply a one-to-many transformation on declaration lists.
This transformation is needed when I implemented the expansion of declaration splices, as a single declaration splice can be expanded into multiple declarations.

`go` is just doing something like `concatMapM` (or `fmap getAp . foldMap . fmap Ap`), based on span information: it applies transformation when the span matched, replace a single declaration with the generated multiple declarations if the transformation succeeded.
At the time I wrote it, I couldn't find any one-to-many grafting function in ghc-exactprint, so I wrote this by hand.

jneira commented:
This close #733
